in_toto/models/metadata.py

```diff
@@ -282,3 +340,120 @@ def _validate_signatures(self):
 
     for signature in self.signatures:
       securesystemslib.formats.ANY_SIGNATURE_SCHEMA.check_match(signature)
+
+
+class AnyMetadata(SerializationMixin, JSONSerializable):
+  """A Metadata Wrapper to provide some common methods."""
+
+  def __init__(self, metadata):
+    self.metadata: Union[Envelope, Metablock] = metadata
+    self.dsse = isinstance(metadata, Envelope)
```

Comment:
I don't think composition is the best suited pattern here, it requires case handling in every function for objects of different classes that could have the same interfaces for serialization, signature creation/verification and payload extraction. 

What if AnyMetadata was a base class for Metablock and Envelope that only implements the `from_dict` factory as you have it now, and defines common interfaces for above tasks? I can see two challenges there. 
1. current signature verification methods expect different arguments for Metablock and Envelope
--> A good solution would be to implement a new additional verify method on Metablock that accepts the same arguments as Envelope's verify method.

2. Envelope is implemented in securesystemslib, so it can't inherit from AnyMetadata in in-toto
--> I can see multiple solutions:
   1. Use protocols to not require subclassing. This would require either dropping 3.7 or adding an additional dependency to pull in the backport
   2. Create a glue Envelope class in in-toto, which could inherit both from AnyMetadata and securesystemslib's Envelope. This solutions seems a bit messy.
   3. Implement AnyMetadata and a Metablock-like class in securesystemslib.  The latter is also requested in [#272](https://github.com/secure-systems-lab/securesystemslib/issues/272). However, this would mean  a breaking change.

What do you think?

Reply:
I found that that it is possible to define a base class `AnyMetadata` for `Envelope` and `Metablock` and use them in it's `from_dict` method. I am surprised to see this behaviour. But if we are going to have a base class `AnyMetadata` then these all metadata need to be present in a single file otherwise they will cause circular import.

It is not possible to do first task without refactoring `Metablock` (see https://github.com/in-toto/in-toto/pull/503#discussion_r954870651) and I am not sure about second task either.